### PR TITLE
Adding UDP logger plugin

### PR DESCRIPTION
**Pull request by sshniro**
The UDP logger sends the logs to a UDP server. 
Fix for #1069

### CODE REVIEW

File: lua/apisix/plugins/udp-logger.lua [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+local ngx = ngx
+
+local timer_at = ngx.timer.at
+local udp = ngx.socket.udp
+
+local schema = {
+    type = "object",
+    properties = {
+        host = {type = "string"},
+        port = {type = "number"},
```
Comment (membphis) [inline lua/apisix/plugins/udp-logger.lua]: It should be 'integer', and the value should more than `0`.

File: lua/apisix/plugins/udp-logger.lua [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        timeout = {type = "number", minimum = 1} --default is 3 seconds
```
Comment (membphis) [inline lua/apisix/plugins/udp-logger.lua]: ditto

Comment (membphis) [reply lua/apisix/plugins/udp-logger.lua]: we can specify the default value directly.

File: lua/apisix/plugins/udp-logger.lua [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    sock:settimeout(conf.timeout)
+    local ok, err = sock:setpeername(conf.host, conf.port)
+
+    if not ok then
+        core.log.error("Failed to connect to UDP server: host[", conf.host, "] port[", conf.port, "] ", err)
+        return
+    end
+
+    ok, err = sock:send(log_message)
+
```
Comment (membphis) [inline lua/apisix/plugins/udp-logger.lua]: please remove this blank line

File: lua/apisix/plugins/udp-logger.lua [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    ok, err = sock:close()
+
+    if not ok then
+        core.log.error("Failed to close the UDP connection, host[", conf.host, "] port[", conf.port, "] ", err)
+    end
+end
+
+
+function _M.log(conf)
+
```
Comment (membphis) [inline lua/apisix/plugins/udp-logger.lua]: delete this blank line

File: lua/apisix/plugins/udp-logger.lua [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    return timer_at(0, log, conf, core.json.encode(log_util.get_full_log(ngx)))
+
```
Comment (membphis) [inline lua/apisix/plugins/udp-logger.lua]: ditto

File: lua/apisix/plugins/udp-logger.lua [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    local sock = udp()
+    sock:settimeout(conf.timeout)
+    local ok, err = sock:setpeername(conf.host, conf.port)
+
+    if not ok then
+        core.log.error("Failed to connect to UDP server: host[", conf.host, "] port[", conf.port, "] ", err)
+        return
+    end
+
+    ok, err = sock:send(log_message)
```
Comment (membphis) [inline lua/apisix/plugins/udp-logger.lua]: I think we'd better add a tail `\r\n` or `\n` for each log message.

**Comment by membphis** [comment]
There are performance issues with the per-request log message mode.

It is better to put log messages in the queue first, and then send them in batches.

**Comment by moonming** [comment]
> There are performance issues with the per-request log message mode.
> 
> It is better to put log messages in the queue first, and then send them in batches.

Done is better then perfect. I think we can use another PR to fix performance issues.

**Comment by membphis** [comment]
```
t/admin/plugins.t ...................... 1/? 
#   Failed test 'TEST 1: get plugins' name - response_body_like - response is expected (["limit-req","limit-count","limit-conn","key-auth","basic-auth","prometheus","node-status","jwt-auth","zipkin","ip-restriction","grpc-transcode","serverless-pre-function","serverless-post-function","openid-connect","proxy-rewrite","redirect","response-rewrite","fault-injection","udp-logger"])'
#   at /home/travis/build/apache/incubator-apisix/test-nginx/lib/Test/Nginx/Socket.pm line 1635.
#                   '["limit-req","limit-count","limit-conn","key-auth","basic-auth","prometheus","node-status","jwt-auth","zipkin","ip-restriction","grpc-transcode","serverless-pre-function","serverless-post-function","openid-connect","proxy-rewrite","redirect","response-rewrite","fault-injection","udp-logger"]
# '
#     doesn't match '(?^:\["limit-req","limit-count","limit-conn","key-auth","prometheus","node-status","jwt-auth","zipkin","ip-restriction","grpc-transcode","serverless-pre-function","serverless-post-function","openid-connect","proxy-rewrite","redirect","response-rewrite","udp-logger"\])'
t/admin/plugins.t ...................... 3/? # Looks like you failed 1 test of 3.
```

File: lua/apisix/plugins/udp-logger.lua [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+    local ok, err = sock:setpeername(conf.host, conf.port)
+    if not ok then
+        core.log.error("Failed to connect to UDP server: host[", conf.host, "] port[", conf.port, "] ", err)
+        return
+    end
+
+    ok, err = sock:send(log_message)
+    if not ok then
+        core.log.error("Failed to send data to UDP server: host[", conf.host, "] port[", conf.port, "] ", err)
```
[the remark this answers is not in the document]
Comment (sshniro) [reply lua/apisix/plugins/udp-logger.lua]: @membphis do you mean to return and end the function after error? In that case don't we have to close the socket? (which is done in the next line of statements)

File: lua/apisix/plugins/udp-logger.lua [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+local _M = {
+    version = 0.1,
+    priority = 400,
+    name = plugin_name,
+    schema = schema,
+}
+
+function _M.check_schema(conf)
+
+    if not conf.timeout then
```
[the remark this answers is not in the document]
Comment (sshniro) [reply lua/apisix/plugins/udp-logger.lua]: aw forgot! will refactor.

File: lua/apisix/plugins/udp-logger.lua (hunk 7 above)
[the remark this answers is not in the document]
Comment (membphis) [reply lua/apisix/plugins/udp-logger.lua]: You are right, sorry for this.

**Comment by sshniro** [comment]
@membphis  The following is the final log sent to the server.

```json
{
  "upstream": "127.0.0.1:80",
  "start_time": 1579231324391,
  "client_ip": "127.0.0.1",
  "service_id": "10",
  "route_id": "5",
  "request": {
    "querystring": {
      "param1": "value"
    },
    "size": 118,
    "uri": "/?param1=value",
    "url": "http://127.0.0.1:9080/?param1=value",
    "headers": {
      "host": "127.0.0.1:9080",
      "apikey": "superSecretAPIKey",
      "accept": "*/*",
      "user-agent": "curl/7.58.0"
    },
    "method": "GET"
  },
  "response": {
    "headers": {
      "accept-ranges": "bytes",
      "content-type": "text/html",
      "date": "Fri, 17 Jan 2020 03:22:04 GMT",
      "server": "APISIX web server",
      "connection": "close",
      "last-modified": "Thu, 29 Aug 2019 23:49:13 GMT",
      "etag": "\"5d686479-289\"",
      "content-length": "649"
    },
    "status": 200,
    "size": 892
  },
  "latency": 0.99992752075195
}
````
should I round the number for latency?

**Comment by membphis** [comment]
> should I round the number for latency?

Recommended to milliseconds.

File: lua/apisix/plugins/log-util.lua [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+    local ctx = ngx.ctx.api_ctx
+    local var = ctx.var
+
+    local url = var.scheme .. "://" .. var.host .. ":" .. var.server_port .. var.request_uri
+
+    local service_name
+    local vars = var
+    if ctx.matched_route and ctx.matched_route.value then
+        service_name = ctx.matched_route.value.desc or
```
[the remark this answers is not in the document]
Comment (sshniro) [reply lua/apisix/plugins/log-util.lua]: added serivice id, and route id as adviced.

**Comment by sshniro** [comment]
@membphis the review comments have been resolved, thanks for the pointers.

**Pull request merged**